sdk/storage/Azure.Storage.DataMovement/src/Shared/DataTransferState.cs

```diff
@@ -111,17 +104,22 @@ public void SetId(string id)
         /// Sets the completion status
         /// </summary>
         /// <param name="status"></param>
-        public async Task SetTransferStatus(StorageTransferStatus status)
+        public Task SetTransferStatus(StorageTransferStatus status)
         {
-            if (!_disposedValue)
+            lock (statusLock)
             {
-                await _statusSemaphore.WaitAsync().ConfigureAwait(false);
                 if (_status != status)
                 {
+                    if (StorageTransferStatus.Completed == status ||
+                        StorageTransferStatus.CompletedWithSkippedTransfers == status ||
+                        StorageTransferStatus.CompletedWithFailedTransfers == status)
+                    {
+                        _completionSource.SetResult(status);
```

Comment:
Should this do a TrySetResult instead? I think this will throw if the task has been cancelled or already completed, which may or may not be the behavior you want here.